In [8]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all" 

from __future__ import print_function

# 4.1 Classifying with Bayesian decision theory

* 이전까지는 정보를 가지고 분류 하다가가 확률을 가지고 분류함 

- 장점 : 소량의 데이터, 여러 개의 분류 항목을 다룸
- 단점 : 입력데이터에 따라 민감함
- 적용 : 명목형 값


* 어디에 속할 확률이 높은가 여부로 판별함
    - if p1(x, y) > p2(x, y), then the class is 1.
    - if p2(x, y) > p1(x, y), then the class is 2.


* Byes?
    * 베이지안확률 이라고하는 분류 사용하기 편하고 인기가 있음 
    * 확실하지 않은 상태의 사전지식과 논리를 적용하는 것이 가능 
    * 빈발 확률 
    * 데이터로부터 결론을 도출해 내고 논리와 사전 지식을 허용하지 않음
    
    
# 4.2 Conditional probability
조건부확률

P(A|B) = p(A∩B)/p(B)<br>
p(A∩B) = p(B|A)*p(A)

    -> P(A|B) = p(B|A)*p(A) / p(B)
    
    
# 4.3 Classifying with conditional probabilities\

If P(c1|x, y) > P(c2|x, y), the class is c1. <br>
If P(c1|x, y) < P(c2|x, y), the class is c2.



# 4.4 Document classification with naïve Bayes


* General approach to naïve Bayes
    1. Collect: Any method. We’ll use RSS feeds in this chapter.
    2. Prepare: Numeric or Boolean values are needed.
    3. Analyze: With many features, plotting features isn’t helpful. Looking at histograms is a better idea.
    4. Train: Calculate the conditional probabilities of the independent features.
    5. Test: Calculate the error rate.
    6. Use: One common application of naïve Bayes is document classification. You can use naïve Bayes in any classification setting. It doesn’t have to be text.



* 가정
    - 모든 feature는 독립적이다. (독립적) 
    - 모든 feature는 중요도가 같다. 



# 4.5 Classifying text with Python

- token : 문자의 조합, URL, IP ..  (feature : 각 단어)

문서에서 속성 (단어사전, 빈도수, 의미 ..) 

token 을 vector 로 만들고 조건부 확률을 계산 <br>
각각의 문서에 대해서 0과 1로 만듬 (폭력적, 비폭력적) <br>
나이브 베이스를 수행하기 위한 분류기 생성 


## 4.5.1 Prepare: making word vectors from text

단어 벡터 : 단어 리스트 


In [3]:
import numpy as np

In [4]:
def loadDataSet():
    postingList=[['my', 'dog', 'has', 'flea', 'problems', 'help', 'please'],
                 ['maybe', 'not', 'take', 'him', 'to', 'dog', 'park', 'stupid'],
                 ['my', 'dalmation', 'is', 'so', 'cute', 'I', 'love', 'him'],
                 ['stop', 'posting', 'stupid', 'worthless', 'garbage'],
                 ['mr', 'licks', 'ate', 'my', 'steak', 'how', 'to', 'stop', 'him'],
                 ['quit', 'buying', 'worthless', 'dog', 'food', 'stupid']]
    classVec = [0,1,0,1,0,1]    #1 is abusive, 0 not
    return postingList,classVec

In [5]:
# 단어 리스트 만들기               
def createVocabList(dataSet):
    # 문서의 단어 목록 
    vocabSet = set([]) # 집합변수 (비어있는 set을 만듬)
    for document in dataSet:
        vocabSet = vocabSet | set(document) 
    return list(vocabSet)

In [7]:
createVocabList(loadDataSet()[0])[:5]

['cute', 'love', 'help', 'garbage', 'quit']

In [9]:
import bayes

In [19]:
def setOfWords2Vec(vocabList, inputSet):
    # [0] 빈 1차원 벡터를 만듬 
    returnVec = [0]*len(vocabList)
    print("\n returnVec : {}".format(returnVec))
    for word in inputSet:
        print("\n word : {}".format(word))
        if word in vocabList:
            print("   vocabList.index : {}".format(vocabList.index(word)))
            returnVec[vocabList.index(word)] = 1
        else: print("the word: %s is not in my Vocabulary!" % word)
        print("   returnVec : {}".format(returnVec))
    return returnVec

In [20]:
postingList,classVec = loadDataSet()
vocabList = createVocabList(postingList)
setOfWords2Vec(vocabList, postingList[0])


 returnVec : [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

 word : my
   vocabList.index : 31
   returnVec : [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]

 word : dog
   vocabList.index : 24
   returnVec : [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1]

 word : has
   vocabList.index : 18
   returnVec : [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1]

 word : flea
   vocabList.index : 10
   returnVec : [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1]

 word : problems
   vocabList.index : 6
   returnVec : [0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1]

 word : help
   vocabList.index : 2
   returnVec : [0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 

[0,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 1]

## 4.5.2 Train: calculating probabilities from word vectors

>p(c|$W$) : w1 이 나왔을때 문서가 폭력적일 확률 

p(c|$W$) = (p($W$|c)p(c)) / p($W$)

* $W$ : 단어 벡터 
* p($W$) : 그 단어나 문서에서 나타날 확률
* p(c) : c1 이 폭력적인 문서일 확률 
* p($W$|c) : 폭력적인 문서에서 해당 단어 W1 이 나올 확률 <br>
      w0, w1.. wn 사건이 독립적이여서 <br>
      p(w0|ci) * p(w1|ci) * .... * p(wn|ci) = p($W$|c) 구할 수 있다. 
    

* 의사 코드 

    분류항목에 대한 문서의 개수 세기 
    for 훈련을 위한 모든 문서 개수 만큼 반복 
        for 분류 항목 개수 만큼 반복
            해당 토큰이 문서 내에 있다면 -> 토큰 개수 증가 
            토큰에 대한 개수 증가 
        for 분류 항목 개수 만큼 반복
            for 토큰의 개수만큼 반복
              조건부 확률를 구하기 위해 해당 토큰의 개수를 전체 토큰 개수로 나눔
    각 분류 항목에 대한 조건부 확률 반환

In [38]:
list_of_post, class_label = bayes.loadDataSet()
my_vocab_list = bayes.createVocabList(list_of_post)

trainMat = []
for post in list_of_post:
    trainMat.append(bayes.setOfWords2Vec(my_vocab_list, post))

In [39]:
trainMat2 = [bayes.setOfWords2Vec(my_vocab_list, post) for post in list_of_post]

In [46]:
trainMat == trainMat2
print(trainMat2[:2])

True

[[0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1], [0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0]]


In [67]:
# 조건부 확률 구하기 (단어의 벡터, 정답 카테고리 )
def trainNB0(trainMatrix, trainCategory):
    numTrainDocs = len(trainMatrix) # 문서 6개 
    numWords = len(trainMatrix[0])  # 단어 32개 
    pAbusive = sum(trainCategory)/float(numTrainDocs) # p(Ci)
    print("prob. of abusive: {}".format(pAbusive))
    
#     p0Num = ones(numWords); p1Num = ones(numWords) # 분자 # 32개 어휘 단어   
#     p0Denom = 2.0; p1Denom = 2.0  # 분모
    p0Num = np.zeros(numWords); p1Num = np.zeros(numWords) # 분자 # 32개 어휘 단어   
    p0Denom = 0.0; p1Denom = 0.0  # 분모
    
    for i in range(numTrainDocs):
        print("\n -- for -- ")
        print("  trainCategory[{}] : {}".format(i, trainCategory[i]))
        # 글이 폭력적일떄
        if trainCategory[i] == 1: 
            #해당단어가 몇번 나왔는지  
            p1Num += trainMatrix[i]
            # 폭력적인 문서의 단어 개수 
            p1Denom += sum(trainMatrix[i])            
            print("    p1 NUM :{} \n    Denom :{}".format(p1Num, p1Denom))
        else:
            p0Num += trainMatrix[i]
            p0Denom += sum(trainMatrix[i])
            print("    p0 NUM :{} \n    Denom :{}".format(p0Num, p0Denom))
    
    # 32개 전체단어중에 단어별로 몇번 등장했는지 (에 로그)
    p1Vect = p1Num/p1Denom #np.log(p1Num/p1Denom) # p(Wi|c1)
    p0Vect = p0Num/p0Denom #np.log(p0Num/p0Denom) # p(Wi|c0)
    print("\n p1Vect : {}".format(p1Vect))
    print("\n p0Vect : {}".format(p0Vect))
    return p0Vect,p1Vect,pAbusive

# 참고) p(Wi|c1) + p(Wi|c0) 한다고 1이 되지 않음 (조건부확률)

In [70]:
# trainMat
# class_label
p0Vect,p1Vect,pAbusive = trainNB0(trainMat, class_label)
p0Vect[26], p1Vect[26]

prob. of abusive: 0.5

 -- for -- 
  trainCategory[0] : 0
    p0 NUM :[ 0.  0.  1.  0.  0.  0.  1.  0.  0.  0.  1.  0.  0.  0.  0.  0.  0.  0.
  1.  0.  0.  0.  0.  1.  1.  0.  0.  0.  0.  0.  0.  1.] 
    Denom :7.0

 -- for -- 
  trainCategory[1] : 1
    p1 NUM :[ 0.  0.  0.  0.  0.  0.  0.  0.  1.  0.  0.  0.  0.  0.  1.  1.  0.  0.
  0.  0.  0.  1.  1.  0.  1.  0.  1.  0.  1.  0.  0.  0.] 
    Denom :8.0

 -- for -- 
  trainCategory[2] : 0
    p0 NUM :[ 1.  1.  1.  0.  0.  1.  1.  1.  0.  0.  1.  1.  0.  0.  0.  1.  0.  0.
  1.  0.  0.  0.  0.  1.  1.  0.  0.  1.  0.  0.  0.  2.] 
    Denom :15.0

 -- for -- 
  trainCategory[3] : 1
    p1 NUM :[ 0.  0.  0.  1.  0.  0.  0.  0.  1.  1.  0.  0.  0.  0.  1.  1.  0.  1.
  0.  1.  0.  1.  1.  0.  1.  0.  2.  0.  1.  0.  0.  0.] 
    Denom :13.0

 -- for -- 
  trainCategory[4] : 0
    p0 NUM :[ 1.  1.  1.  0.  0.  1.  1.  1.  0.  1.  1.  1.  1.  0.  0.  2.  0.  0.
  1.  0.  1.  1.  0.  1.  1.  1.  0.  1.  0.  1.  1.  3.] 
    Denom :24.0


(0.0, 0.15789473684210525)

* -- for -- <br>
trainCategory[5] : 1 <br>
    p1 NUM :[ 0.  0.  0.  1.  1.  0.  0.  0.  1.  1.  0.  0.  0.  1.  1.  1.  1.  1. 0.  2.  0.  1.  1.  0.  2.  0.  3.  0.  1.  0.  0.  0.] <br>
    Denom :19.0

확률로 폭력적인 문장에서 나오는 단어들 19개 (2번 나온 단어, 3번 나온 단어 stupid)


* 
p1Vect : [ 0.          0.          0.          0.05263158  0.05263158  0.          0.  0.          0.05263158  0.05263158  0.          0.          0.  0.05263158  0.05263158  0.05263158  0.05263158  0.05263158  0.   0.10526316  0.         0.05263158  0.05263158  0.          0.10526316  0.          0.15789474  0.          0.05263158  0.          0.          0.        ]
  
폭력적인 문서에서 1번째 단어가 나올 확률은 p(W1|c1)=0

stupid 단어가 각 문서에서 확률 (0.0, 0.15789473684210525)

pAbusive : p(C1)

* p( $W$ ) 분모는 어짜피 비교하는 값이라 같은건 구할필요 없어서 구하지 않음


## 4.5.3 Test: modifying the classifier for real-world conditions

실행활에 맞추기 위해서 분류기를 수정함

* trainNB0 <br>
확률이 0인 경우가 있어서 확률을 곱하면 전체 확률이 0가 됨 <br>
모든 단어가 적어도 한번은 나오는 것을 가정해서 확률이 0인것을 없게 함
> p0Num = ones(numWords); p1Num = ones(numWords) # 분자 # 32개 어휘 단어 

    모든단어를 1로 초기화했을 때 입력받은 문장이 하나도 없는 경우 <br>
    모든 1/2 가 되어서 두가지 분류 기준에 어디든 들어갈수 있게 만듬 
> p0Denom = 2.0; p1Denom = 2.0  # 다른단어에 영향을 주지 않고 영향력을 줄임 


* 언더플로우 : 작은 숫자끼리 너무 많이 곱해서 발행하는 문제 (0.1 * 0.000002 * ...)

** 이러한 문제를 없애기 위해 log 를 씌움 **
- 큰 스케일 값을 작게 표현하기 위해서 ($log_{10} 10=1$, $log_{10} 100=2$)
- 곱샘이 -> 덧샘으로 변하게 되어 계산이 편해짐 (dB 데시벨, 지진규모 ..)
- 증감의 범위나 분포가 바뀌지 않음
    > p1Vect = np.log(p1Num/p1Denom) # p(Wi|c1)<br>
    p0Vect = np.log(p0Num/p0Denom) # p(Wi|c0)

In [71]:
reload(bayes)

<module 'bayes' from 'bayes.py'>

In [72]:
# 변경된 train 함수 
p0Vect,p1Vect,pAbusive = bayes.trainNB0(trainMat, class_label)
p0Vect[26], p1Vect[26]

(-3.2580965380214821, -1.6582280766035324)

In [81]:
# 검증 (분류하고 싶은 문서의 단어벡터, 폭력적일때 단어확률, 비폭력적일때 단어확률, 폭력적인 문석일확률)
def classifyNB(vec2Classify, p0Vec, p1Vec, pClass1):
    # log 값들의 연산이라 * -> + 으로 연산함 
    # sum(vec2Classify * p1Vec) == p(w0|ci) * p(w1|ci) * .... * p(wn|ci) = p($W$|c)
    # log(pClass1) == p(C1)
    p1 = sum(vec2Classify * p1Vec) + np.log(pClass1)    
    p0 = sum(vec2Classify * p0Vec) + np.log(1.0 - pClass1)
    
    print("\nvec2Classify :{}".format(vec2Classify))
    print("\np1Vec :{}".format(p1Vec))
    print("\np0Vec :{}".format(p0Vec))
    print("\npClass1 :{}".format(pClass1))
    print("\np1 :{} p0 :{}".format(p1,p0))
    
    # p(C1|W) > p(C0|W)
    if p1 > p0:
        return 1
    else: 
        return 0

In [93]:
p0v,p1v,pAbus = bayes.trainNB0(trainMat, class_label)
testEntry =  ['I', 'dumped', 'garbage', 'for', 'my', 'cute', 'dog']
testVec = bayes.setOfWords2Vec(my_vocab_list, testEntry)
classifyNB(testVec, p0v, p1v, pAbus)

the word: dumped is not in my Vocabulary!
the word: for is not in my Vocabulary!

vec2Classify :[1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1]

p1Vec :[-3.04452244 -3.04452244 -3.04452244 -2.35137526 -2.35137526 -3.04452244
 -3.04452244 -3.04452244 -2.35137526 -2.35137526 -3.04452244 -3.04452244
 -3.04452244 -2.35137526 -2.35137526 -2.35137526 -2.35137526 -2.35137526
 -3.04452244 -1.94591015 -3.04452244 -2.35137526 -2.35137526 -3.04452244
 -1.94591015 -3.04452244 -1.65822808 -3.04452244 -2.35137526 -3.04452244
 -3.04452244 -3.04452244]

p0Vec :[-2.56494936 -2.56494936 -2.56494936 -3.25809654 -3.25809654 -2.56494936
 -2.56494936 -2.56494936 -3.25809654 -2.56494936 -2.56494936 -2.56494936
 -2.56494936 -3.25809654 -3.25809654 -2.15948425 -3.25809654 -3.25809654
 -2.56494936 -3.25809654 -2.56494936 -2.56494936 -3.25809654 -2.56494936
 -2.56494936 -2.56494936 -3.25809654 -2.56494936 -3.25809654 -2.56494936
 -2.56494936 -1.87180218]

pClass1 

0

In [91]:
def testingNB():
    listOPosts,listClasses = loadDataSet()
    myVocabList = bayes.createVocabList(listOPosts)
    trainMat=[]
    for postinDoc in listOPosts:
        trainMat.append(bayes.setOfWords2Vec(myVocabList, postinDoc))
    p0V,p1V,pAb = bayes.trainNB0(np.array(trainMat),np.array(listClasses))
    testEntry = ['love', 'my', 'dalmation']
    thisDoc = np.array(bayes.setOfWords2Vec(myVocabList, testEntry))
    print(testEntry,'classified as: ', bayes.classifyNB(thisDoc,p0V,p1V,pAb))
    testEntry = ['stupid', 'garbage']
    thisDoc = np.array(bayes.setOfWords2Vec(myVocabList, testEntry))
    print(testEntry,'classified as: ', bayes.classifyNB(thisDoc,p0V,p1V,pAb))

testingNB()

['love', 'my', 'dalmation'] classified as:  0
['stupid', 'garbage'] classified as:  1


## 4.5.4 Prepare: the bag-of-words document model


중복단어모델 (bag Of Words)

> setOfWords2Vec : 단어가 여러번 나와도 한번으로 치부함 (단어의집합)

> bagOfWords2VecMN : 중복되어 나온 단어의 값을 + 여 처리 

단어가 여러번 나오면 중요도가 높다고 생각 할 때


In [ ]:
def bagOfWords2VecMN(vocabList, inputSet):
    returnVec = [0]*len(vocabList)
    for word in inputSet:
        if word in vocabList:
            returnVec[vocabList.index(word)] += 1 #<- 요기 다름 
    return returnVec

def setOfWords2Vec(vocabList, inputSet):
    returnVec = [0]*len(vocabList)
    for word in inputSet:
        if word in vocabList:
            returnVec[vocabList.index(word)] = 1
    return returnVec

# 4.6 Example: classifying spam email with naïve Bayes

## 4.6.1 Prepare: tokenizing text


In [96]:
mySent='This book is the best book on Python or M.L. I have ever laid eyes upon.'

# 정규표현식을 이용 하여 문장 -> 단어 분류 
# \W : white space (공백, ',', '.', \n, \r, \t..)
def textParse(bigString):    
    import re
    listOfTokens = re.split(r'\W*', bigString)  # white space
    return [tok.lower() for tok in listOfTokens if len(tok) > 2] # 2글자 이상, 소문자로 변경 
textParse(mySent)

['this',
 'book',
 'the',
 'best',
 'book',
 'python',
 'have',
 'ever',
 'laid',
 'eyes',
 'upon']

In [100]:
emailText = open('data/ch04/email/ham/6.txt').read()
textParse(emailText)

['hello',
 'since',
 'you',
 'are',
 'owner',
 'least',
 'one',
 'google',
 'groups',
 'group',
 'that',
 'uses',
 'the',
 'customized',
 'welcome',
 'message',
 'pages',
 'files',
 'are',
 'writing',
 'inform',
 'you',
 'that',
 'will',
 'longer',
 'supporting',
 'these',
 'features',
 'starting',
 'february',
 '2011',
 'made',
 'this',
 'decision',
 'that',
 'can',
 'focus',
 'improving',
 'the',
 'core',
 'functionalities',
 'google',
 'groups',
 'mailing',
 'lists',
 'and',
 'forum',
 'discussions',
 'instead',
 'these',
 'features',
 'encourage',
 'you',
 'use',
 'products',
 'that',
 'are',
 'designed',
 'specifically',
 'for',
 'file',
 'storage',
 'and',
 'page',
 'creation',
 'such',
 'google',
 'docs',
 'and',
 'google',
 'sites',
 'for',
 'example',
 'you',
 'can',
 'easily',
 'create',
 'your',
 'pages',
 'google',
 'sites',
 'and',
 'share',
 'the',
 'site',
 'http',
 'www',
 'google',
 'com',
 'support',
 'sites',
 'bin',
 'answer',
 'answer',
 '174623',
 'with',
 'the',


## 4.6.2 Test: cross validation with naïve Bayes


In [123]:
# 스팸 분류기 
def spamTest():
    docList=[]; classList = []; fullText =[]
    # 인스턴스 25개 읽기
    for i in range(1,26):
        # 바로 분류 시킴 
        wordList = bayes.textParse(open('data/ch04/email/spam/%d.txt' % i).read())
        # 배열로 추가함
        docList.append(wordList)
        # 한문장으로 모든 스팸 인트턴스 넣음 
        fullText.extend(wordList)
        # 스팸
        classList.append(1)
        
        wordList = bayes.textParse(open('data/ch04/email/ham/%d.txt' % i).read())
        docList.append(wordList)
        fullText.extend(wordList)
        classList.append(0)
    
    # docList 이차원 배열 
    vocabList = bayes.createVocabList(docList) # 단어사전 만들기 
    trainingSet = range(50); testSet=[]  #0~49 까지 index 리스트 
    
    # test 에 사용할 인스턴스 추출
    for i in range(10):
        # uniform : 균등분포로 난수 발생 (int 로 정수화시킴)
        randIndex = int(np.random.uniform(0,len(trainingSet))) 
        testSet.append(trainingSet[randIndex])
        # test 데이터로 선정된 값 삭제
        del(trainingSet[randIndex])  
    
    trainMat=[]; trainClasses = []
    # trainingSet : training data 의 index 값 
    for docIndex in trainingSet:
        # docIndex 모든 이메일에 있는 단어들 리스트 
        # 단어가 나온 횟수만큼 count
        trainMat.append(bayes.bagOfWords2VecMN(vocabList, docList[docIndex]))
        trainClasses.append(classList[docIndex])
      
    p0V,p1V,pSpam = bayes.trainNB0(np.array(trainMat),np.array(trainClasses))
    print("p0v: {} \np1v: {} \nspam: {}\n".format(p0V[:10],p1V[:10],pSpam))
    
    errorCount = 0
    for docIndex in testSet:        #classify the remaining items
        wordVector = bayes.bagOfWords2VecMN(vocabList, docList[docIndex])
        if bayes.classifyNB(np.array(wordVector),p0V,p1V,pSpam) != classList[docIndex]:
            errorCount += 1
            print("classification error ",docList[docIndex])
            print("correct answer ", classList[docIndex])
    print('the error rate is: ',float(errorCount)/len(testSet))
    #return vocabList,fullText


In [110]:
# 원래 다른 값인데 int 하면서 같은 값이 됨
[int(np.random.uniform(0,50)) for _ in range(10)] 

[22, 49, 31, 12, 10, 33, 11, 49, 39, 23]

In [126]:
reload(bayes)
spamTest()

<module 'bayes' from 'bayes.pyc'>

p0v: [-5.31811999 -6.70441435 -6.01126717 -6.01126717 -6.70441435 -6.70441435
 -6.01126717 -6.01126717 -6.01126717 -6.70441435] 
p1v: [-4.95053137 -6.33682573 -6.33682573 -6.33682573 -4.95053137 -5.64367855
 -6.33682573 -6.33682573 -6.33682573 -4.13960115] 
spam: 0.45

classification error  ['yay', 'you', 'both', 'doing', 'fine', 'working', 'mba', 'design', 'strategy', 'cca', 'top', 'art', 'school', 'new', 'program', 'focusing', 'more', 'right', 'brained', 'creative', 'and', 'strategic', 'approach', 'management', 'the', 'way', 'done', 'today']
correct answer  0
classification error  ['scifinance', 'now', 'automatically', 'generates', 'gpu', 'enabled', 'pricing', 'risk', 'model', 'source', 'code', 'that', 'runs', '300x', 'faster', 'than', 'serial', 'code', 'using', 'new', 'nvidia', 'fermi', 'class', 'tesla', 'series', 'gpu', 'scifinance', 'derivatives', 'pricing', 'and', 'risk', 'model', 'development', 'tool', 'that', 'automatically', 'generates', 'and', 'gpu', 'enabled', 'source', 'cod

spam 메일의 확률에 error rate 가 크게 영향을 받진 않았다.

신규 데이터가 들어오면 신규데이터가 추가된 모든 데이터를 다시 학습해야한다. 

    >> Knowledge Representation : p0V, p1V, pSpam 
    
참고) 신경망에서는 노드들 간의 연결정보를 조절하는데 훈련된 데이터를 다음 epoch 에 사용해도 큰 영향을 주지 않는다. (cross-validataion)



# 4.7 Example: using naïve Bayes to reveal local attitudes from personal ads

* RSS LIST 사이트 <br>
https://newyork.craigslist.org/

개인광고에 포함된 지역의 특색 도출 

## 4.7.1 Collect: importing RSS feeds

* rss feed parser 설치 <br>
    http://code.google.com/p/feedparser/ <br>
    (C:\Anaconda3) C:\Users\student>conda install -n ml27 feedparser


* stop word (중지단어) : usually the most common words in a language, there is no single universal list of stop words used by all natural language processing tools, and indeed not all tools even use such a list.

In [129]:
import feedparser
ny = feedparser.parse("https://newyork.craigslist.org/search/stp?format=rss")
len(ny)
ny['entries'][0]['summary']

11

u"average guy seeking a cool female friend that I can chat with laugh bug out and just be ourselves around each other. Someone who has a good head on a shoulders that's funny mature and cool to hang with it doesnt matter what borough you're from just b ..."

In [ ]:
# list 에서 가장 많이 나온 상위 30 개만 
def calcMostFreq(vocabList,fullText, num):
    import operator
    freqDict = {}
    for token in vocabList:
        freqDict[token]=fullText.count(token)
    
    #value 값으로 정렬 내림차순 
    sortedFreq = sorted(freqDict.iteritems(), key=operator.itemgetter(1), reverse=True) #py3 freqDict
    return sortedFreq[:num]       

In [130]:
def localWords(feed1,feed0):
    import feedparser
    docList=[]; classList = []; fullText =[]
    minLen = min(len(feed1['entries']),len(feed0['entries']))
    
    # 두개를 비교하기 위해 최소값로 비교 
    for i in range(minLen):
        # 요약 정보만 
        wordList = textParse(feed1['entries'][i]['summary'])
        docList.append(wordList)
        fullText.extend(wordList)
        classList.append(1) #NY is class 1
        wordList = textParse(feed0['entries'][i]['summary'])
        docList.append(wordList)
        fullText.extend(wordList)
        classList.append(0)
        
    # 모든 단어 리스트 
    vocabList = bayes.createVocabList(docList)#create vocabulary
    # 가장 빈번하게 나온것 
    top30Words = calcMostFreq(vocabList,fullText)   #remove top 30 words
    
    # 상위 30 개의 단어를 단어리스트에서 제거 (for, you, the, ... )
    for pairW in top30Words:
        # 사전의 pairW = key, value
        if pairW[0] in vocabList: vocabList.remove(pairW[0])
    
    # ny 30 개, sf 25 개에서 각각 25개씩 50개 가져옴
    trainingSet = range(2*minLen); testSet=[]           #create test set
    
    # test set 20개
    for i in range(20):
        randIndex = int(np.random.uniform(0,len(trainingSet)))
        testSet.append(trainingSet[randIndex])
        del(trainingSet[randIndex])  
    
    trainMat=[]; trainClasses = []
    # training matrix 만들기 
    for docIndex in trainingSet:#train the classifier (get probs) trainNB0
        trainMat.append(bayes.bagOfWords2VecMN(vocabList, docList[docIndex]))
        trainClasses.append(classList[docIndex])
    
    # training 시키기 pSpam : 뉴욕인 확률 
    p0V,p1V,pSpam = bayes.trainNB0(np.array(trainMat),np.array(trainClasses))
    
    errorCount = 0
    for docIndex in testSet:        #classify the remaining items
        wordVector = bayes.bagOfWords2VecMN(vocabList, docList[docIndex])
        if bayes.classifyNB(np.array(wordVector),p0V,p1V,pSpam) != classList[docIndex]:
            errorCount += 1
            print('the error is: ',docList[docIndex])
            print('the correct is: ',classList[docIndex])
    print('the error rate is: ',float(errorCount)/len(testSet))
    return vocabList,p0V,p1V

In [133]:
reload(bayes)
ny = feedparser.parse("https://newyork.craigslist.org/search/stp?format=rss")
sf = feedparser.parse("https://sfbay.craigslist.org/search/stp?format=rss")
vocabList, pSF, pNY = bayes.localWords(ny, sf)
len(vocabList)
pSF[:10]
pNY[:10]

<module 'bayes' from 'bayes.pyc'>

('the error rate is: ', 0.3)


557

array([-5.6347896 , -5.6347896 , -4.94164242, -5.6347896 , -5.6347896 ,
       -4.94164242, -5.6347896 , -5.6347896 , -5.6347896 , -5.6347896 ])

array([-5.86646806, -5.86646806, -5.86646806, -5.86646806, -5.86646806,
       -5.86646806, -4.76785577, -5.86646806, -4.4801737 , -4.76785577])

## 4.7.2 Analyze: displaying locally used words

지역적으로 사용된 단어 표시 



In [138]:
def getTopWords(ny,sf):
    import operator
    vocabList,p0V,p1V=bayes.localWords(ny,sf)
    topNY=[]; topSF=[]
    for i in range(len(p0V)):
        if p0V[i] > -6.0 : topSF.append((vocabList[i],p0V[i]))
        if p1V[i] > -6.0 : topNY.append((vocabList[i],p1V[i]))
    sortedSF = sorted(topSF, key=lambda pair: pair[1], reverse=True)
    print("SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**")
    for item in sortedSF:
        print(item[0])
    sortedNY = sorted(topNY, key=lambda pair: pair[1], reverse=True)
    print("NY**NY**NY**NY**NY**NY**NY**NY**NY**NY**NY**NY**NY**NY**NY**NY**")
    for item in sortedNY:
        print(item[0])

In [139]:
getTopWords(ny, sf)

('the error rate is: ', 0.3)
SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**
long
friend
male
doing
think
make
safe
lady
females
women
orral
want
pleasure
good
got
sensual
face
clean
please
only
feels
bay
come
many
blonde
wants
things
other
host
drama
range
your
drive
interested
newbie
asian
skin
passion
conversating
friendly
woman
confounding
drinking
accidentally
discrete
enjoy
hanging
what
giving
while
lights
new
here
hours
wherever
family
prefer
husbands
strrings
takes
lickking
few
high
room
hour
this
ride
movies
meet
minimum
lacks
sit
hou
possibly
beauty
may
such
maybe
wine
hobby
into
they
now
society
event
special
receiving
free
struggle
teach
definition
times
place
eatting
feel
thick
little
families
their
white
foreplay
hotel
believe
herself
than
boyfriend
matter
thhongs
loves
look
take
seductive
sure
serviced
most
but
don
disease
judgmental
find
busy
going
black
pretty
his
photography
cannot
married
sex
individual
ways
pictures
gentleman
horny
email
taking
etc
